### PR TITLE
Refactor remote pull to provide progress

### LINKED ISSUE #7543: Podman remote commands do not provide same textual feedback as local
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**
When running podman remotely against a Fedora 32 VM, the commands don't appear to return as much information as running locally.

**Steps to reproduce the issue:**

1. brew install podman, config remote podman according to https://www.redhat.com/sysadmin/podman-clients-macos-windows
2. run a podman command, such as podman pull. no images are stored on the system.

**Describe the results you received:**
```
[cwawak@hulk containers]$ podman --log-level debug pull nginx
INFO[0000] podman filtering at log level debug
DEBU[0000] Called pull.PersistentPreRunE(podman --log-level debug pull nginx)
DEBU[0000] public key signer enabled for identity "/Users/cwawak/.ssh/ovirt"
DEBU[0000] Found SSH_AUTH_SOCK "/private/tmp/com.apple.launchd.XY3CFb4BBX/Listeners", ssh-agent signer enabled
DEBU[0000] public key signer enabled for identity "/Users/cwawak/.ssh/ovirt"
DEBU[0000] Found SSH_AUTH_SOCK "/private/tmp/com.apple.launchd.XY3CFb4BBX/Listeners", ssh-agent signer enabled
INFO[0000] Setting parallel job count to 25
4bb46517cac397bdb0bab6eba09b0e1f8e90ddd17cf99662997c3253531136f8
DEBU[0009] Called pull.PersistentPostRunE(podman --log-level debug pull nginx)
```

**Describe the results you expected:**
this is directly on the server:
```
$ podman pull nginx
Trying to pull registry.fedoraproject.org/nginx...
  manifest unknown: manifest unknown
Trying to pull registry.access.redhat.com/nginx...
  unsupported: This repo requires terms acceptance and is only available on registry.redhat.io
Trying to pull registry.centos.org/nginx...
  manifest unknown: manifest unknown
Trying to pull docker.io/library/nginx...
Getting image source signatures
Copying blob b49ea07d2e93 done
Copying blob bf5952930446 done
Copying blob cb9a6de05e5a [======================================] 25.2MiB / 25.2MiB
Copying blob a5e4a503d449 done
Copying blob 9513ea0afb93 done
Copying config 4bb46517ca done
Writing manifest to image destination
Storing signatures
4bb46517cac397bdb0bab6eba09b0e1f8e90ddd17cf99662997c3253531136f8
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version` from client:**

```
$ podman version
Client:
Version:      2.0.6
API Version:  1
Go Version:   go1.15
Built:        Wed Sep  2 13:30:31 2020
OS/Arch:      darwin/amd64

Server:
Version:      2.0.5
API Version:  0
Go Version:   go1.14.6
Built:        Wed Dec 31 19:00:00 1969
OS/Arch:      linux/amd64
```

**Output of `podman info --debug` from client:**

```
host:
  arch: amd64
  buildahVersion: 1.15.1
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.19-1.fc32.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.19, commit: 5dce9767526ed27f177a8fa3f281889ad509fea7'
  cpus: 1
  distribution:
    distribution: fedora
    version: "32"
  eventLogger: file
  hostname: podman
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.8.4-200.fc32.x86_64
  linkmode: dynamic
  memFree: 2375626752
  memTotal: 4119506944
  ociRuntime:
    name: crun
    package: crun-0.14.1-4.fc32.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-1.fc32.x86_64
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 2
  swapFree: 4254068736
  swapTotal: 4254068736
  uptime: 19m 13.82s
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /home/cwawak/.config/containers/storage.conf
  containerStore:
    number: 6
    paused: 0
    running: 1
    stopped: 5
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.2-1.fc32.x86_64
      Version: |-
        fusermount3 version: 3.9.1
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.1
        using FUSE kernel interface version 7.31
  graphRoot: /home/cwawak/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 4
  runRoot: /run/user/1000/containers
  volumePath: /home/cwawak/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Wed Dec 31 19:00:00 1969
  GitCommit: ""
  GoVersion: go1.14.6
  OsArch: linux/amd64
  Version: 2.0.5
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
client: $ brew info podman
podman: stable 2.0.6 (bottled)

server: $ rpm -q podman
podman-2.0.5-1.fc32.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Server running a fedora 32 minimal image.

### PR DESCRIPTION
**Pull request by jwhonce**
Fixes #7543

Tested with:

$ (echo '# start'; podman-remote pull nginx ) 2>&1 | ts '[%Y-%m-%d %H:%M:%.S]'
$ (echo '# start'; podman pull nginx ) 2>&1 | ts '[%Y-%m-%d %H:%M:%.S]'

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by jwhonce** [comment]
@vrothberg This moves remote closer to abi, but the progress bars are still off.  Do you have experience with why they would be different?

**Comment by edsantiago** [comment]
@jwhonce you need to rebase on master to fix CI. My fault, sorry.

File: cmd/podman/images/pull.go [hunk 1]
```diff
@@ -77,21 +77,18 @@ func init() {
...
 	flags.BoolVarP(&pullOptions.Quiet, "quiet", "q", false, "Suppress output information when pulling images")
 	flags.StringVar(&pullOptions.SignaturePolicy, "signature-policy", "", "`Pathname` of signature policy file (not usually used)")
-	flags.BoolVar(&pullOptions.TLSVerifyCLI, "tls-verify", true, "Require HTTPS and verify certificates when contacting registries")
 
-	if registry.IsRemote() {
-		_ = flags.MarkHidden("authfile")
-		_ = flags.MarkHidden("cert-dir")
-		_ = flags.MarkHidden("tls-verify")
+	if !registry.IsRemote() {
+		flags.StringVar(&pullOptions.Authfile, "authfile", auth.GetDefaultAuthFile(), "Path of the authentication file. Use REGISTRY_AUTH_FILE environment variable to override")
```
Comment (vrothberg) [inline cmd/podman/images/pull.go:89]: Unrelated to this PR: I believe pull can/should support --authfile.

**Comment by vrothberg** [comment]
> @vrothberg This moves remote closer to abi, but the progress bars are still off. Do you have experience with why they would be different?

I'll fetch the PR and have a look

**Comment by vrothberg** [comment]
> > @vrothberg This moves remote closer to abi, but the progress bars are still off. Do you have experience with why they would be different?
> 
> I'll fetch the PR and have a look

Got it. We have a check in containers/image to not render the progress bars if the provided writer is not a TTY. The reasoning was to prevent verbose and ugly logs. In such cases, we only print a single line indicating an operation.

I will play a bit with c/image and see what we can do. I see it as an extension to this PR that we can tackle later.

File: pkg/api/handlers/libpod/images_pull.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+				authfile,
+				writer,
+				&dockerRegistryOptions,
+				image.SigningOptions{},
+				nil,
+				util.PullImageAlways)
+			if err != nil {
+				stderr.Write([]byte(err.Error() + "\n"))
+			}
+			images = append(images, newImage.ID())
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/images_pull.go:144]: Nil deref: We should only append if `err == nil`.

File: pkg/bindings/images/pull.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	stderr := io.Writer(os.Stderr)
+	if options.Quiet {
+		stderr = ioutil.Discard
+	}
+
+	dec := json.NewDecoder(response.Body)
+	var mErr error
+	for {
```
Comment (vrothberg) [inline pkg/bindings/images/pull.go:67]: We need to check if the connection has been lost or closed by the server. I killed the server and the client continued waiting.

File: pkg/api/handlers/libpod/images_pull.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Header().Add("Content-Type", "application/json")
+	flush()
+
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(true)
+	var failed bool
+loop:
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/images_pull.go:161]: Could you drop a comment why this label is needed? I had to search and guess that others may not know as well.

File: pkg/api/handlers/libpod/images_pull.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			flush()
+		case <-runCtx.Done():
+			if !failed {
+				report.Images = images
+				if err := enc.Encode(report); err != nil {
+					logrus.Warnf("Failed to json encode error %q", err.Error())
+				}
+				flush()
+			}
+			break loop
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/images_pull.go:186]: Same here: // Break outer loop?

File: pkg/bindings/images/pull.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+		switch {
+		case report.Stream != "":
+			fmt.Fprint(stderr, report.Stream)
+		case report.Error != "":
+			mErr = multierror.Append(mErr, errors.New(report.Error))
+		case len(report.Images) > 0:
+			return report.Images, nil
+		default:
+			return nil, errors.New("failed to parse pull results stream, unexpected input")
```
Comment (vrothberg) [inline pkg/bindings/images/pull.go:84]: Can we you include the `report`?

File: pkg/bindings/images/pull.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	if !response.IsSuccess() {
+		return nil, response.Process(err)
+	}
+
+	stderr := io.Writer(os.Stderr)
```
Comment (vrothberg) [inline pkg/bindings/images/pull.go:60]: Can you drop a comment that the pull output goes to stderr? I was a bit confused at first but it's correct as is.

File: pkg/domain/entities/images.go [hunk 8]
```diff
@@ -150,7 +150,9 @@ type ImagePullOptions struct {
 
 // ImagePullReport is the response from pulling one or more images.
 type ImagePullReport struct {
-	Images []string
+	Stream string   `json:"stream,omitempty"`
```
Comment (vrothberg) [inline pkg/domain/entities/images.go:153]: Please add comments.

**Review by vrothberg — COMMENTED** [review]
Just a couple of nits.

Note that rendering progress bars as the local client could be a longer journey:
* The progress bars are rendered by containers/image (i.e., the backend) and we don't have a notion of a front end.
* We could hijack the connection as we do with attach and allow to control the client terminal.
* Or we do some extensive rewrite in c/image, and allow the rendering on the client side (via progress messages from the server).

I think keeping the output as is may be the best option at the moment. Hijacking seems overly complex and I don't think we have enough resources to break the progress bars out of c/image.

**Review by vrothberg — COMMENTED** [review]
LGTM, nice work!

**Comment by vrothberg** [comment]
Restarted what looked like a flake in the ubuntu job

**Pull request merged**